packages/repository/src/relations/has-many/has-many-through-repository.factory.ts

```diff
@@ -0,0 +1,219 @@
+ */
+export function createHasManyThroughRepositoryFactory<
+  Target extends Entity,
+  TargetID,
+  Through extends Entity,
+  ThroughID,
+  ForeignKeyType
+>(
+  relationMetadata: HasManyThroughDefinition,
+  targetRepositoryGetter: Getter<EntityCrudRepository<Target, TargetID>>,
+  throughRepositoryGetter: Getter<EntityCrudRepository<Through, ThroughID>>,
+): HasManyThroughRepositoryFactory<Target, Through, ForeignKeyType> {
+  const meta = resolveHasManyThroughMetadata(relationMetadata);
+  debug('Resolved HasMany relation metadata: %o', meta);
+  return function(fkValue?: ForeignKeyType) {
+    async function getAdvancedConstraint(
+      targetInstance: Target,
+    ): Promise<AdvancedConstraint<Target | Through>> {
+      return createAdvancedConstraint<
+        Target,
+        Through,
+        ThroughID,
+        ForeignKeyType
+      >(targetInstance, meta, throughRepositoryGetter, fkValue);
+    }
+    return new DefaultHasManyThroughRepository<
+      Target,
+      TargetID,
+      EntityCrudRepository<Target, TargetID>,
+      Through,
+      ThroughID,
+      EntityCrudRepository<Through, ThroughID>
+    >(targetRepositoryGetter, throughRepositoryGetter, getAdvancedConstraint);
+  };
+}
+
+type HasManyThroughResolvedDefinition = HasManyThroughDefinition & {
+  keyTo: string;
+  targetFkName: string;
+  targetPrimaryKey: string;
+};
+
+async function createAdvancedConstraint<
+  Target extends Entity,
+  Through extends Entity,
+  ThroughID,
+  ForeignKeyType
+>(
+  targetInstance: Target,
+  meta: HasManyThroughResolvedDefinition,
+  throughRepositoryGetter: Getter<EntityCrudRepository<Through, ThroughID>>,
+  fkValue?: ForeignKeyType,
+): Promise<AdvancedConstraint<Target | Through>> {
+  // tslint:disable-next-line:no-any
+  const advancedConstraint: any = {
+    dataObject: {[meta.keyTo]: fkValue} as DataObject<Through>,
+    filter: {},
+    where: {},
+  };
+  if (targetInstance) {
+    // through constraint
+    advancedConstraint.dataObject[meta.targetFkName] =
+      targetInstance[meta.targetPrimaryKey as keyof Target];
+    advancedConstraint.where = {where: advancedConstraint.dataObject};
+  } else {
+    // target constraint
+    const throughRepo = await throughRepositoryGetter();
+    const throughInstances = await throughRepo.find(
+      constrainFilter(undefined, advancedConstraint.dataObject),
+    );
+    if (!throughInstances.length) {
+      const id =
+        'through constraint ' + JSON.stringify(advancedConstraint.dataObject);
+      throw new EntityNotFoundError(throughRepo.entityClass, id);
+    }
+    advancedConstraint.dataObject = {};
+    advancedConstraint.where = {
+      or: throughInstances.map((throughInstance: Through) => {
+        return {id: throughInstance[meta.targetFkName as keyof Through]};
+      }),
+    };
+  }
+  advancedConstraint.filter = {
+    where: advancedConstraint.where,
+  };
+  return advancedConstraint as AdvancedConstraint<Target | Through>;
```

Comment:
I find this weird. What model (Target or Through) is this object constraining? How are the consumers supposed to know that?

Maybe the `AdvancedConstraint` type need more template arguments, e.g. `interface AdvancedConstraint<Through, Target> {...}`?

Reply:
Refer to the previous comment

Reply:
Again, this is now irrelevant with the constraint refactor.

https://github.com/strongloop/loopback-next/pull/2359#issuecomment-475979639